metrics_test.go

```diff
@@ -0,0 +1,203 @@
+
+		lsm_metric := expvar.Get("badger_v4_written_to_l0")
+		require.Equal(t, expectedSize*int64(num), lsm_metric.(*expvar.Int).Value())
+
+		compactionMetric := expvar.Get("badger_v4_compaction_written_bytes").(*expvar.Map)
+		require.Equal(t, nil, compactionMetric.Get("l6"))
+
+		// Force compaction
+		db.Close()
+
+		_, err := OpenManaged(opt)
+		require.NoError(t, err)
+
+		compactionMetric = expvar.Get("badger_v4_compaction_written_bytes").(*expvar.Map)
+		require.GreaterOrEqual(t, expectedSize*int64(num)+int64(num*200), compactionMetric.Get("l6").(*expvar.Int).Value())
+		// Because we have random values, compression is not able to do much, so we incur a cost on total size
+	})
+}
+
+func TestVlogMetrics(t *testing.T) {
+	opt := getTestOptions("")
+	opt.managedTxns = true
+	opt.CompactL0OnClose = true
+	runBadgerTest(t, &opt, func(t *testing.T, db *DB) {
+		ClearAllMetrics()
+		num := 10
+		val := make([]byte, 1<<20) // Large Value
+		key := make([]byte, 40)
+		for i := 0; i < num; i++ {
+			_, err := rand.Read(key)
+			require.NoError(t, err)
+			_, err = rand.Read(val)
+			require.NoError(t, err)
+
+			writer := db.NewManagedWriteBatch()
+			require.NoError(t, writer.SetEntryAt(NewEntry(key, val), 1))
+			writer.Flush()
+		}
+
+		expectedSize := int64(len(val)) + 200 // vlog expected size
+
+		totalWrites := expvar.Get("badger_v4_vlog_writes_num")
+		require.Equal(t, int64(num), totalWrites.(*expvar.Int).Value())
+
+		bytesWritten := expvar.Get("badger_v4_vlog_written_bytes")
+		require.GreaterOrEqual(t, expectedSize*int64(num), bytesWritten.(*expvar.Int).Value())
+
+		txn := db.NewTransactionAt(2, false)
+		item, err := txn.Get(key)
+		require.NoError(t, err)
+		require.Equal(t, uint64(1), item.Version())
+
+		err = item.Value(func(val []byte) error {
+			totalReads := expvar.Get("badger_v4_vlog_reads_num")
+			bytesRead := expvar.Get("badger_v4_vlog_reads_bytes")
+			require.Equal(t, int64(1), totalReads.(*expvar.Int).Value())
+			require.GreaterOrEqual(t, expectedSize, bytesRead.(*expvar.Int).Value())
+			return nil
+		})
+
+		require.NoError(t, err)
+	})
+}
+
+func TestReadMetrics(t *testing.T) {
+	opt := getTestOptions("")
+	opt.managedTxns = true
+	opt.CompactL0OnClose = true
+	runBadgerTest(t, &opt, func(t *testing.T, db *DB) {
+		ClearAllMetrics()
+		num := 10
+		val := make([]byte, 1<<15)
+		keys := [][]byte{}
+		writer := db.NewManagedWriteBatch()
+		for i := 0; i < num; i++ {
+			keyB := key("byte", 1)
+			keys = append(keys, []byte(keyB))
+
+			_, err := rand.Read(val)
+			require.NoError(t, err)
+
+			require.NoError(t, writer.SetEntryAt(NewEntry([]byte(keyB), val), 1))
+		}
+		writer.Flush()
+
+		txn := db.NewTransactionAt(2, false)
+		item, err := txn.Get(keys[0])
+		require.NoError(t, err)
+		require.Equal(t, uint64(1), item.Version())
+
+		totalGets := expvar.Get("badger_v4_gets_total")
+		require.Equal(t, int64(1), totalGets.(*expvar.Int).Value())
+
+		totalMemtableReads := expvar.Get("badger_v4_memtable_gets_total")
```

Comment:
should this be named, badger_v4_gets_total_memtable & the next one as badger_v4_gets_total_lsm_level?